leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/QueuedRequestSender.java

```diff
@@ -0,0 +1,282 @@
+ * request is enqueued first and then sent if a client is back online. If the client does not support the queue mode in
+ * its binding, then the delegate sender is called in order to send the request.
+ */
+public class QueuedRequestSender implements LwM2mRequestSender, Stoppable {
+    private static final Logger LOG = LoggerFactory.getLogger(QueuedRequestSender.class);
+    private final LwM2mRequestSender delegateSender;
+    private final ExecutorService processingExecutor = new ExceptionAwareExecutorService(
+            Executors.newCachedThreadPool(new NamedThreadFactory("leshan-qmode-processingExecutor-%d")));
+    private final Map<Long, ResponseContext> responseContextHolder = new ConcurrentHashMap<>();
+    private final MessageStore messageStore;
+    private final QueuedRequestFactory queuedRequestFactory;
+    private final QueueModeClientRegistryListener queueModeClientRegistryListener;
+    private final QueueModeObservationRegistryListener queueModeObservationRegistryListener;
+    private final ClientRegistry clientRegistry;
+    private final ObservationRegistry observationRegistry;
+    private final ClientStatusTracker clientStatusTracker;
+
+    /**
+     * Creates a new QueueRequestSender using given builder.
+     *
+     * @param builder sender builder
+     */
+    private QueuedRequestSender(Builder builder) {
+        this.messageStore = builder.messageStore;
+        this.queuedRequestFactory = builder.queuedRequestFactory;
+        this.clientRegistry = builder.clientRegistry;
+        this.observationRegistry = builder.observationRegistry;
+        this.delegateSender = builder.delegateSender;
+
+        this.clientStatusTracker = new ClientStatusTracker();
+
+        this.queueModeClientRegistryListener = new QueueModeClientRegistryListener();
+        clientRegistry.addListener(queueModeClientRegistryListener);
+        this.queueModeObservationRegistryListener = new QueueModeObservationRegistryListener();
+        observationRegistry.addListener(queueModeObservationRegistryListener);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public <T extends LwM2mResponse> T send(Client destination, DownlinkRequest<T> request, Long requestTimeout)
+            throws InterruptedException {
+        // Send without callbacks are intended to be synchronous calls and will not be queued.
+        // using the delegate sender to send it immediately.
+        return delegateSender.send(destination, request, requestTimeout);
+    }
+
+    @Override
+    public <T extends LwM2mResponse> void send(Client destination, DownlinkRequest<T> request,
+            final ResponseCallback<T> responseCallback, final ErrorCallback errorCallback) {
+
+        LOG.trace("send()");
+        // safe because DownlinkRequest does not use generics itself
+        @SuppressWarnings("unchecked")
```

Comment:
it seems generics and Queue mode are not really good friends.. and this will not be better with the new "send" API I proposed before :/